File: android/src/main/java/com/batch/batch_flutter/BatchPluginConfiguration.java
package com.batch.batch_flutter;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


/**
 * Manages Batch configuration for the flutter plugin.
 * Configuration will be read from the manifest, but can be overridden at app launch.
 */
public class BatchPluginConfiguration {

    private static final String APIKEY_MANIFEST_KEY = "com.batch.flutter.apikey";
    private static final String PROFILE_CUSTOM_ID_MIGRATION_ENABLED_MANIFEST_KEY = "com.batch.flutter.profile_custom_id_migration_enabled";
    private static final String PROFILE_CUSTOM_DATA_MIGRATION_ENABLED_MANIFEST_KEY = "com.batch.flutter.profile_custom_data_migration_enabled";
    private static final String INITIAL_DND_STATE_MANIFEST_KEY = "com.batch.flutter.do_not_disturb_initial_state";

    private boolean didReadManifest = false;

    @Nullable
    private String apiKey;
    private boolean initialDoNotDisturbState = false;
    private boolean profileCustomIdMigrationEnabled = true;
    private boolean profileCustomDataMigrationEnabled = true;

    synchronized void initFromManifest(@NonNull Context context) {
        //noinspection ConstantConditions
        if (context == null) {
            return;
        }
        if (didReadManifest) {
            return;
        }
        didReadManifest = true;

        final ManifestReader manifestReader = new ManifestReader(context);
        apiKey = manifestReader.readString(APIKEY_MANIFEST_KEY, null);
        initialDoNotDisturbState = manifestReader.readBoolean(INITIAL_DND_STATE_MANIFEST_KEY, false);
        profileCustomIdMigrationEnabled =  manifestReader.readBoolean(PROFILE_CUSTOM_ID_MIGRATION_ENABLED_MANIFEST_KEY, true);
        profileCustomDataMigrationEnabled =  manifestReader.readBoolean(PROFILE_CUSTOM_DATA_MIGRATION_ENABLED_MANIFEST_KEY, true);
    }
    //region Public API

    /**
     * Get the API key.
     *
     * @return The API key.
     */
    @Nullable
    public String getApiKey() {
        return apiKey;
    }

    /**
     * Set the API key.
     *
     * @param apiKey The API key.
     * @return This BatchPluginConfiguration instance for method chaining.
     */
    public BatchPluginConfiguration setAPIKey(@Nullable String apiKey) {
        this.apiKey = apiKey;
        return this;
    }

    /**
     * Get the initial do not disturb state.
     *
     * @return The initial do not disturb state.
     */
    public boolean getInitialDoNotDisturbState() {
        return initialDoNotDisturbState;
    }

    /**
     * Set the initial do not disturb state.
     *
     * @param initialDoNotDisturbState The initial do not disturb state.
     * @return This BatchPluginConfiguration instance for method chaining.
     */
    public BatchPluginConfiguration setInitialDoNotDisturbState(boolean initialDoNotDisturbState) {
        this.initialDoNotDisturbState = initialDoNotDisturbState;
        return this;
    }

    /**
     * Whether custom id migration is enabled or not.
     */
    public boolean isProfileCustomIdMigrationEnabled() {
        return profileCustomIdMigrationEnabled;
    }

    /**
     * Whether Batch should automatically identify logged-in user when running the SDK for the first time.
     * <p>
     * This mean user with a custom_user_id will be automatically attached a to a Profile and could be targeted within a Project scope.
     * @param profileCustomIdMigrationEnabled whether custom id migration is enabled or not.
     * @return This BatchPluginConfiguration instance for method chaining.
     */
    public BatchPluginConfiguration setProfileCustomIdMigrationEnabled(boolean profileCustomIdMigrationEnabled) {
        this.profileCustomIdMigrationEnabled = profileCustomIdMigrationEnabled;
        return this;
    }

    /**
     * Whether custom data migration is enabled or not.
     */
    public boolean isProfileCustomDataMigrationEnabled() {
        return profileCustomDataMigrationEnabled;
    }

    /**
     *  Set whether Batch should automatically attach current installation's data (language/region/customDataAttributes...)
     *  to the User's Profile when running the SDK for the first time.
     *
     * @param profileCustomDataMigrationEnabled whether custom data migration is enabled or not.
     * @return This BatchPluginConfiguration instance for method chaining.
     */
    public BatchPluginConfiguration setProfileCustomDataMigrationEnabled(boolean profileCustomDataMigrationEnabled) {
        this.profileCustomDataMigrationEnabled = profileCustomDataMigrationEnabled;
        return this;
    }
    //endregion
}

File: ios/batch_flutter/Sources/batch_flutter/BatchPluginConfiguration.swift
import Foundation
import Batch

fileprivate struct PlistKeys {
    static let APIKey = "BatchFlutterAPIKey"
    static let initialDnDState = "BatchFlutterDoNotDisturbInitialState"
    static let profileCustomIdMigrationEnabled = "BatchFlutterProfileCustomIdMigrationEnabled"
    static let profileCustomDataMigrationEnabled = "BatchFlutterProfileCustomDataMigrationEnabled"
}


/**
 Manages Batch's Flutter plugin configuration. Do not instantiate this directly, use BatchFlutterPlugin to get an instance: your changes will otherwise be ignored.
 
 This class' default values are initialized using your Info.plist settings, if any.
 */
@objc
public class BatchPluginConfiguration: NSObject {
    
    /// The Batch API Key
    public var APIKey: String? = nil

    /// The initial do not disturb state.
    public var initialDoNotDisturbState: Bool = false
    
    /// Whether Batch should automatically identify logged-in user when running the SDK for the first time.
    public var profileCustomIdMigrationEnabled: Bool = true
    
    /// Whether Batch should automatically attach current installation's data (language/region/customDataAttributes...) to the User's Profile when running the SDK for the first time.
    public var profileCustomDataMigrationEnabled: Bool = true
    
    /// Store the API Key in a different variable, so that editing "APIKey" has no effect until "apply" is called
    internal var actualAPIKey: String? = nil
    
    /// Flag indicating if this configuration has already been setup from the info.plist file
    private var didReadInfoPlist = false
    
    override init() {
        super.init()
        readFromPlist()
    }
    
    internal func readFromPlist() {
        if (didReadInfoPlist) {
            return
        }
        didReadInfoPlist = true
        
        APIKey = PlistReader.readString(key: PlistKeys.APIKey)
        initialDoNotDisturbState = PlistReader.readBoolean(key: PlistKeys.initialDnDState, fallbackValue: initialDoNotDisturbState)
        profileCustomIdMigrationEnabled = PlistReader.readBoolean(key: PlistKeys.profileCustomIdMigrationEnabled, fallbackValue: true)
        profileCustomDataMigrationEnabled = PlistReader.readBoolean(key: PlistKeys.profileCustomDataMigrationEnabled, fallbackValue: true)
    }
    
    /// Apply configuration. Returns true on success.
    internal func apply() -> Bool {
        if let APIKey = APIKey, APIKey.trimmingCharacters(in: .whitespacesAndNewlines).count > 0 {
            actualAPIKey = APIKey
            BatchMessaging.doNotDisturb = initialDoNotDisturbState
            var migrations : BatchMigration = []
            if (!profileCustomIdMigrationEnabled) {
                BatchFlutterLogger.logDebug(module: "Plugin", message: "Disabling profile custom id migration.")
                migrations.insert(.customID)
            }
            if (!profileCustomDataMigrationEnabled) {
                BatchFlutterLogger.logDebug(module: "Plugin", message: "Disabling profile custom data migration.")
                migrations.insert(.customData)
            }
            BatchSDK.setDisabledMigrations(migrations)
            return true
        } else {
            // No API key
            return false
        }
    }
}
